packages/airnode-protocol/contracts/WIP/Airnode.sol

```diff
@@ -0,0 +1,273 @@
+        bytes32 templateId,
+        address reporter,
+        address sponsor,
+        address sponsorWallet,
+        address fulfillAddress,
+        bytes4 fulfillFunctionId,
+        bytes calldata parameters
+    ) external returns (bytes32 requestId) {
+        require(fulfillAddress != address(this), "Fulfill address AirnodeRrp");
+        require(
+            sponsorToRequesterToSponsorshipStatus[sponsor][msg.sender],
+            "Requester not sponsored"
+        );
+        uint256 requesterRequestCount = requesterToRequestCountPlusOne[
+            msg.sender
+        ];
+        requestId = keccak256(
+            abi.encodePacked(
+                block.chainid,
+                address(this),
+                msg.sender,
+                requesterRequestCount,
+                templateId,
+                reporter,
+                sponsor,
+                sponsorWallet,
+                fulfillAddress,
+                fulfillFunctionId,
+                parameters
+            )
+        );
+        requestIdToFulfillmentParameters[requestId] = keccak256(
+            abi.encodePacked(
+                templates[templateId].airnode,
+                reporter,
+                sponsorWallet,
+                fulfillAddress,
+                fulfillFunctionId
+            )
+        );
+        requesterToRequestCountPlusOne[msg.sender]++;
+        emit MadeRequest(
+            reporter,
+            requestId,
+            requesterRequestCount,
+            block.chainid,
+            msg.sender,
+            templateId,
+            sponsor,
+            sponsorWallet,
+            fulfillAddress,
+            fulfillFunctionId,
+            parameters
+        );
+    }
+
+    function fulfillRequest(
+        bytes32 requestId,
+        address airnode,
+        address reporter,
+        address fulfillAddress,
+        bytes4 fulfillFunctionId,
+        bytes calldata data,
+        bytes calldata signature
+    ) external returns (bool callSuccess, bytes memory callData) {
+        require(
+            keccak256(
+                abi.encodePacked(
+                    airnode,
+                    reporter,
+                    msg.sender,
+                    fulfillAddress,
+                    fulfillFunctionId
+                )
+            ) == requestIdToFulfillmentParameters[requestId],
+            "Invalid request fulfillment"
+        );
+        require(
+            (
+                keccak256(abi.encodePacked(requestId, data))
+                    .toEthSignedMessageHash()
+            ).recover(signature) == airnode,
+            "Invalid signature"
+        );
+        delete requestIdToFulfillmentParameters[requestId];
+        (callSuccess, callData) = fulfillAddress.call( // solhint-disable-line avoid-low-level-calls
+            abi.encodeWithSelector(fulfillFunctionId, requestId, data)
+        );
+        if (callSuccess) {
+            emit FulfilledRequest(reporter, requestId, data);
```

Comment:
Again, we're emitting `reporter` instead of `airnode` because it's the `reporter` listening for this event